### ISSUE TITLE
Perfect shuffle without buffer of size len(dataset)

### ISSUE THREAD

**Issue by DUPASYoann** [tensorflow/tensorflow#53242]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.7.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
The new feature will allow perfect shuffle without using a buffer of size len(dataset). When using tf.data.Dataset object for data pipeline, to faster the data process during model training, we use tf.data.Dataset.cache method. But, with large dataset or/and limited RAM capacity, we cannot use perfect shuffle method because it need the same amount of RAM capacity and can cause an Out of Memory error. 

***Exemple of actual behaviour***

```python
ds = tf.data.Dataset.range(100000)  # large dataset
ds = ds.map(SomeTransformation)
ds = ds.cache() # RAM is 70% use
ds = ds.map(SomeTransformationThatUseRandomBehavior)
ds = ds.shuffle(buffer_size=len(ds))  # (Perfect) shuffle cannot be done because of an OOM Error 

...

model.fit(ds)
```

***Exemple of the new feature***

```python
ds = tf.data.Dataset.range(100000)  # large dataset
ds = ds.map(SomeTransformation)
ds = ds.cache() # RAM is 70% use
ds = ds.map(SomeTransformationThatUseRandomBehavior)
ds = ds.shuffle(buffer_size=None)  # (Perfect) shuffle will not cause an OOM Error because 
                                   # it will no longer need a buffer to work

...

model.fit(ds)
```

***Current solution but not perfect***

```python
ds = tf.data.Dataset.range(100000)  # large dataset
ds = ds.map(SomeTransformation)
ds = ds.cache("path/to/file") # RAM is no longer use
ds = ds.map(SomeTransformationThatUseRandomBehavior)
ds = ds.shuffle(buffer_size=len(ds))  # (Perfect) shuffle work but take more time
                                      # to read from a cache file than read from a cache  in RAM

...

model.fit(ds)
```

**Will this change the current api? How?**
tf.data.Dataset.shuffle now allow buffer_size to be set to None

**Who will benefit with this feature?**
 Anyone who are using large datasets and those who are limited with RAM capacity (can only fit once the data in RAM)

**Other Info**
An another shuffle option, if perfect shuffle cannot be done due to some technical constraint, can be to randomly select an element of the dataset. that's mean 2 elements can be draw.

**Comment by sanatmpa1** [comment]
@DUPASYoann,

Can you please elaborate you feature request by adding more details and also can you provide an example if possible? Thanks!

**Comment by DUPASYoann** [comment]
> @DUPASYoann,
> 
> Can you please elaborate you feature request by adding more details and also can you provide an example if possible? Thanks!

Edit done !

**Comment by sachinprasadhs** [comment]
It is normally suggested not to use `cache` for large dataset since it occupies your RAM usage, you can still proceed without using `cache` and then use shuffle which can be performed since your RAM usage will be less.
If you are concerned about the speed, you can do it in other ways, either you can shuffle the dataset at source(OS level) and then create tf.data and cache or to shuffle in other ways you can split the source data into multiple files and read them in the random order during training using `tf.data.Dataset.list_files`

**Comment by DUPASYoann** [comment]
> It is normally suggested not to use cache for large dataset since it occupies your RAM usage, you can still proceed without using cache and then use shuffle which can be performed since your RAM usage will be less.

Yeah, I agree, and that what I do with `tf.data.Dataset.cache(filename=filepath)`. In this way, i can use perfect `shuffle` with `reshuffle_each_iteration=True` since cache is no longer on RAM. But perfect `shuffle` can be only ensure if you can store all your dataset in RAM since you have to set the shuffle buffer to len(dataset). 

> you can shuffle the dataset at source(OS level) and then create tf.data and cache

For me it's not an ideal solution, because when you use `cache` operation on your dataset, the order will be "frozen" and so you will only benefit the shuffle once. (When i say the order will be "frozen", I mean that in every epoch (If I consider one epoch equal one entire loop over my dataset) I will have the same order).

> split the source data into multiple files and read them in the random order during training using tf.data.Dataset.list_files

This is the best solution right now. But consider this case : 

#### Actual situation

```python
ds = tf.data.Dataset.list_files("path/to/files")  # few elements but each element are large image
ds = ds.map(SomeTransformation)
ds = ds.cache() # cache the dataset because the transformation before take long time 
                # and we don't have to do it multiple time since the result are the same
ds = ds.map(SomeTransformationThatUseRandomBehaviorLikeRandomCropOnImage)
ds = ds.shuffle(buffer_size=len(ds))  # Perfect shuffle can only be ensure if we have enough RAM to store all elements

...

model.fit(ds) # It's important that we have random order every epoch
```

#### Solution when RAM is not enough

```python 
ds = tf.data.Dataset.list_files("path/to/files")  # few elements but elements are large image
ds = ds.map(SomeTransformation)

for data in ds:
    pickle.dump(data, "path/for/element{i}") # cache data with our own method 

new_ds = tf.data.Dataset.list_files("path/off/all/elements/*", shuffle=True) # we don't need to use tf.data.Dataset.shuffle anymore
new_ds = new_ds.map(lambda x: tf.numpy_function(read_data_with_pickle_load, [x], [ds.elements_spec.dtype])
new_ds = new_ds.map(SomeTransformationThatUseRandomBehaviorLikeRandomCropOnImage)


...

model.fit(ds) # We still have a random order every epoch
```

On further consideration, the new feature can be a shuffle argument on `tf.data.Dataset.cache` like the argument on `tf.data.Dataset.list_files`. When True, It would allow random shuffle of the cached dataset (file or RAM). 

#### With the new feature

```python
ds = tf.data.Dataset.list_files("path/to/files")  # few elements but elements are large image
ds = ds.map(SomeTransformation)
ds = ds.cache(shuffle=True) # we can now easily have perfect shuffle, and if the dataset is cached in RAM,
                            # it's now only store once
ds = ds.map(SomeTransformationThatUseRandomBehaviorLikeRandomCropOnImage)

...

model.fit(ds) # we would still have our random order every epoch
̀```

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
